### PR TITLE
Fix non-azure openai generation issues - part 1

### PR DESCRIPTION
**Pull request by MaryGao**
this is part of the issue https://github.com/Azure/autorest.typescript/issues/2042 and fixes https://github.com/Azure/autorest.typescript/issues/2046.

- Support cutom http auth in Modular and enable auth testing in cadl-ranch
- Fix error response issue not from azure core
- Fallback to package name if no scope defined in package name
- Add a pipeline property in classical client to allow customizing policy

### CODE REVIEW

File: packages/typespec-ts/src/transform/transformSchemas.ts [hunk 1]
```diff
@@ -131,7 +131,7 @@ export function transformSchemas(
           (!program.stateMap(modelKey).get(prop[1].type) ||
             !program.stateMap(modelKey).get(prop[1].type)?.includes(context))
         ) {
-          if (prop[1].type.name === "Error") {
+          if (isAzureCoreErrorType(prop[1].type)) {
```
Comment (MaryGao) [inline packages/typespec-ts/src/transform/transformSchemas.ts:134]: Only skip the error model when it is from azure core namespace.

File: packages/typespec-ts/test/modularIntegration/authOauth2.spec.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import {
+  createOAuth2,
+  invalid,
+  OAuth2Context,
+  valid
+} from "./generated/authentication/oauth2/src/api/index.js";
+import { assert } from "chai";
+import { customBearerTokenAuthenticationPolicy } from "../util/customBearerTokenTestingPolicy.js";
+
+describe("OAuth2Context in API Layer", () => {
```
[the remark this answers is not in the document]
Comment (qiaozha) [reply packages/typespec-ts/test/modularIntegration/authOauth2.spec.ts:14]: Based on our past experience when handling the customer issues for mgmt plane, we do recommend our customers to use customized policy for serveral times. Personally, I don't think it worth a breaking here to recommend them to custom it in the api layer.  

If we want to have an exact same user experience like mgmt plane to use `client.pipeline.addPolicy()` and `client.pipeline.removePolicy()` , we have to add a new property in the classical client layer.
`public pipeline: Pipeline;`  and change the constructor like 
```typescript
export class ApiKeyClient {
  private _client: ApiKeyContext;
  public pipeline: Pipeline;

  /** Illustrates clients generated with ApiKey authentication. */
  constructor(credential: KeyCredential, options: ApiKeyClientOptions = {}) {
    this._client = createApiKey(credential, options);
    this.pipeline = this._client.pipeline;
  }
  ...
}
```
Not sure if others have different opinions or have other options about it.  @joheredi thoughts ?

[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/authOauth2.spec.ts:14]: I prefer to have a `pipeline` in classical client and we could refactor these logic to core whever possible in future.

[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/authOauth2.spec.ts:14]: updated in commit https://github.com/Azure/autorest.typescript/pull/2041/commits/ab2632ba96a7bc4c2f4ef2032040e39910e24847

[the remark this answers is not in the document]
Comment (qiaozha) [reply packages/typespec-ts/test/modularIntegration/authOauth2.spec.ts:14]: why not change the test into using the classical client layer pipeline ?

**Review by qiaozha — APPROVED** [review]

File: packages/typespec-ts/test/modularIntegration/authOauth2.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaryGao) [reply packages/typespec-ts/test/modularIntegration/authOauth2.spec.ts:14]: https://github.com/Azure/autorest.typescript/blob/637821101ed96dd7cc8981ba9f02ee4f44ddfa77/packages/typespec-ts/test/modularIntegration/authUnion.spec.ts#L77

**Comment by MaryGao** [comment]
@joheredi I merged this pr but feel free to add any comment if you have!

**Pull request merged**